### ISSUE TITLE
ConPTY inserts unexpected new line character

### ISSUE THREAD

**Issue by MaxRis** [Microsoft/console#380]
* Microsoft Windows [Version 10.0.17763.316] ( also reproducible on 10.0.18841.1000 )

* What you're doing and what's happening:
[ConPTY MiniTerm sample](https://github.com/Microsoft/console/tree/master/samples/ConPTY/MiniTerm) receives extra new line character to display `tmux` bottom status bar.
![2019-03-03_21-15-50-miniterm-tmux-extranewline](https://user-images.githubusercontent.com/1928902/53700318-a8397680-3df9-11e9-9ed2-e9d42721917b.png)

* What's wrong / what should be happening instead:
ConPTY should process input VT sequences without inserting new line character.
The screenshot of the same input, but inside MSYS2 terminal emulator:
![2019-03-03_21-20-52-msys2-renders-tmux-bottom-line-correctly](https://user-images.githubusercontent.com/1928902/53700379-5cd39800-3dfa-11e9-9ba7-16f076726e20.png)

* Steps to reproduce:
    - Build console CPP app to read input data from file and send to standard output 
[main.txt](https://github.com/Microsoft/console/files/2923572/main.txt)
    - File with input data [input.txt](https://github.com/Microsoft/console/files/2923565/input.txt) to be placed in the same directory with console app executable from previous step
   - Build adjusted version of MiniTerm https://github.com/MaxRis/console/commit/82f36fd39b4fc7e12752d226dd5d70c0b3159824 , run it, launch executable from first step.

* Additional info:
    - Similar issue is reproducible with ConEmu. Extra new line is added after `tmux` status line:
![2019-03-03_21-50-09-conemu-tmux-statusline-rendering-issue](https://user-images.githubusercontent.com/1928902/53700761-74148480-3dfe-11e9-9b97-b52eff24ede6.png)
    - Following conclusion made on top of following observations might be important:
       Changing [SGR command](https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_(Select_Graphic_Rendition)_parameters) length has straightforward impact on position where the extra new line character is generated by ConPTY
![2019-03-03_22-04-08-sgr-3-symbols](https://user-images.githubusercontent.com/1928902/53701208-57c71680-3e03-11e9-9d05-73d4bad65737.png)
![2019-03-03_22-05-53-sgr-4-symbols](https://user-images.githubusercontent.com/1928902/53701213-5c8bca80-3e03-11e9-800d-9137b6d871b7.png)
![2019-03-03_22-10-47-sgr-18-symbols](https://user-images.githubusercontent.com/1928902/53701217-61e91500-3e03-11e9-8146-83766afbd614.png)
       ConPTY might has an issue to process SGR command and count symbols related to the SGR command as output terminal emulator buffer symbols. Since the length of line is 130 symbols in above case, ConPTY inserts new line character every time when total count of SGR command symbols and visible output buffer symbols exceeds 130 symbols.
       Here it's simplified input data used to generate above output with SGR command of different length 
[input.txt](https://github.com/Microsoft/console/files/2923644/input.txt)

**Comment by DHowett-MSFT** [comment]
Hey, about this "ConsolePrinter": is it turning on `ENABLE_VIRTUAL_TERMINAL_PROCESSING`?

If it _isn't_, something fun is happening with your VT sequences. When VT processing is disabled, those VT sequences are being treated as raw characters to display in the console. Since those characters _are_ being displayed, they're counted towards the width of the output.

When you're connected over ConPTY, two strange (but by-design) things happen:

1. Those raw escape sequences are line-wrapped, because you didn't tell conhost that it should process them
2. The ConPTY consumer receives those sequences _and then parses them._ Since they may have been broken by the line wrapping ro any other console screen buffer manipulation, they _will_ be rendered incorrectly.

In short, if you're not setting `ENABLE_VIRTUAL_TERMINAL_OUTPUT` but you are in fact using virtual terminal output, this is by design.

**Comment by MaxRis** [comment]
Hey @DHowett-MSFT thank you for looking into this!

`ENABLE_VIRTUAL_TERMINAL_PROCESSING` is set for MiniTerm.exe https://github.com/Microsoft/console/blob/master/samples/ConPTY/MiniTerm/MiniTerm/Terminal.cs#L36 and I have believed that the console instance and custom set attributes are shared with launched `CMD.exe` and further with `ConsolePrinter.exe` process. Does it make any sense here?

`ConsolePrinter` doesn't modify console mode and just writes to standard output:
```
#include <iostream>
#include <fstream>
#include <string>

void main() {
	std::ifstream dataFile("input.txt");
	if (dataFile.is_open()) {
		std::string dataLine;
		while (std::getline(dataFile, dataLine)) {
			std::cout << dataLine;
		}
	}
}
```

**Comment by DHowett-MSFT** [comment]
Ah, so the code in MiniTerm is enabling VT processing _for the terminal itself_. Since ConsolePrinter is on the other side:

```
+-------+                      +-------+
| Mini  | <----------(pipe)--- | pty   |
|  Term | -----------(pipe)--> |  host | - vt processing OFF.
+-------+                      +-------+
  |   ^                           |  ^  
  v   |                           v  |
+-------+                      +---------+
| con   |                      | Console |
|  host | - VT processing on.  | Printer |
+-------+                      +---------+
  |   ^ 
  v   | 
/ - - - +
: user  :
:       :
+ - - - /
```

it needs to enable virtual terminal processing _for its console host_.

**Comment by MaxRis** [comment]
@DHowett-MSFT thanks! will give it a try!

**Comment by MaxRis** [comment]
@DHowett-MSFT thank you for finding out the source of issue!

If I'm extending `ConsolePrinter` with something like following:
```
        HANDLE hOutput = GetStdHandle(STD_OUTPUT_HANDLE);
	DWORD dwMode;
	GetConsoleMode(hOutput, &dwMode);
	dwMode |= ENABLE_PROCESSED_OUTPUT | ENABLE_VIRTUAL_TERMINAL_PROCESSING;
	if (!SetConsoleMode(hOutput, dwMode)) {
		exit(1);
	}
```
Everything is rendered correctly, even under ConEmu! Thanks!

**Comment by oising** [comment]
> Ah, so the code in MiniTerm is enabling VT processing _for the terminal itself_. Since ConsolePrinter is on the other side:
> 
> ```
> +-------+                      +-------+
> | Mini  | <----------(pipe)--- | pty   |
> |  Term | -----------(pipe)--> |  host | - vt processing OFF.
> +-------+                      +-------+
>   |   ^                           |  ^  
>   v   |                           v  |
> +-------+                      +---------+
> | con   |                      | Console |
> |  host | - VT processing on.  | Printer |
> +-------+                      +---------+
>   |   ^ 
>   v   | 
> / - - - +
> : user  :
> :       :
> + - - - /
> ```
> it needs to enable virtual terminal processing _for its console host_.

These are the kinds of notes that belong on Docs. 

> Hey, about this "ConsolePrinter": is it turning on `ENABLE_VIRTUAL_TERMINAL_PROCESSING`?
> 
> If it _isn't_, something fun is happening with your VT sequences. When VT processing is disabled, those VT sequences are being treated as raw characters to display in the console. Since those characters _are_ being displayed, they're counted towards the width of the output.
> 
> When you're connected over ConPTY, two strange (but by-design) things happen:
> 
> 1. Those raw escape sequences are line-wrapped, because you didn't tell conhost that it should process them
> 2. The ConPTY consumer receives those sequences _and then parses them._ Since they may have been broken by the line wrapping ro any other console screen buffer manipulation, they _will_ be rendered incorrectly.
> 
> In short, if you're not setting `ENABLE_VIRTUAL_TERMINAL_OUTPUT` but you are in fact using virtual terminal output, this is by design.

This is the kind of information that belongs on Docs, along with all of the PseudoConsole stuff.

**Comment by miniksa** [comment]
Please file a docs bug at https://github.com/MicrosoftDocs/Console-Docs. Feel free to make a branch and add suggested text to a pull request if you have a good idea of what you think it should say.

**Comment by Biswa96** [comment]
Inspired from @DHowett-MSFT's ascii art. Does ConPTY work with Duplex Named Pipe? As following:

```
+-------+                      +-------+
| Mini  |                      | pty   |
|  Term | <-------(pipe)-----> |  host | - vt processing OFF.
+-------+                      +-------+
    ^                              ^  
    |                              |
    v                              v  
+-------+                      +---------+
| con   |                      | Console |
|  host | - VT processing on.  | Printer |
+-------+                      +---------+
    ^ 
    | 
    v
/ - - - +
: user  :
:       :
+ - - - /
```

**Comment by DHowett-MSFT** [comment]
Give it a shot!
